contracts/ethregistrar/ETHRegistrarController.sol

```diff
@@ -3,176 +3,312 @@ pragma solidity >=0.8.4;
-
-        uint expires;
-        if(resolver != address(0)) {
-            // Set this contract as the (temporary) owner, giving it
-            // permission to set up the resolver.
-            expires = base.register(tokenId, address(this), duration);
-
-            // The nodehash of this label
-            bytes32 nodehash = keccak256(abi.encodePacked(base.baseNode(), label));
-
-            // Set the resolver
-            base.ens().setResolver(nodehash, resolver);
-
-            // Configure the resolver
-            if (addr != address(0)) {
-                Resolver(resolver).setAddr(nodehash, addr);
-            }
-
-            // Now transfer full ownership to the expeceted owner
-            base.reclaim(tokenId, owner);
-            base.transferFrom(address(this), owner, tokenId);
-        } else {
-            require(addr == address(0));
-            expires = base.register(tokenId, owner, duration);
+        uint256 duration = prices.duration(
+            name,
+            base.nameExpires(uint256(label)),
+            msg.value
+        );
+        Cost memory cost = _consumeCommitment(
+            name,
+            duration,
+            makeCommitment(
+                name,
+                owner,
+                secret,
+                resolver,
+                data,
+                reverseRecord,
+                fuses
+            )
+        );
+
+        uint256 expires = nameWrapper.registerAndWrapETH2LD(
+            name,
+            owner,
+            duration,
+            resolver,
+            fuses
+        );
+
+        if (data.length > 0) {
+            require(
+                resolver != address(0),
+                "ETHRegistrarController: resolver is required when data is supplied"
+            );
+            _setRecords(resolver, label, data);
         }
 
-        emit NameRegistered(name, label, owner, cost, expires);
-
-        // Refund any extra payment
-        if(msg.value > cost) {
-            payable(msg.sender).transfer(msg.value - cost);
+        if (reverseRecord) {
+            _setReverseRecord(name, resolver, msg.sender);
         }
+
+        emit NameRegistered(
+            name,
+            label,
+            owner,
+            cost.base,
+            cost.premium,
+            expires
+        );
     }
 
-    function renew(string calldata name, uint duration) external payable {
-        uint cost = rentPrice(name, duration);
-        require(msg.value >= cost);
+    function renew(string calldata name, uint256 duration) external payable {
+        Cost memory cost = rentPrice(name, duration);
+        require(msg.value >= cost.base);
 
         bytes32 label = keccak256(bytes(name));
-        uint expires = base.renew(uint256(label), duration);
+        uint256 expires = base.renew(uint256(label), duration);
 
-        if(msg.value > cost) {
-            payable(msg.sender).transfer(msg.value - cost);
+        if (msg.value > cost.base) {
```

Comment:
Shouldn't we do the same as for registration, and just consume whatever was sent?